### PR TITLE
ref(reprocessing2): Make reprocessing more crash resilient

### PR DESCRIPTION
**Pull request by untitaker**
* reprocess_event is now inlined into reprocess_group. That comes with a large performance penalty but allows you to catch exceptions happening during setup of the processing task, such as missing attachments, missing payloads, or something else entirely going wrong. In all those cases you can simply skip over processing the event and instead do the default action (keep/delete).
* reprocess_event now checks whether any plugins require attachments to be present before kicking off the task. If attachments are missing the event will instead just be moved/deleted. This allows you to reprocess an entire group where only the first n events have attachments without surprising side effects.
* events without a reprocessing payload are no longer sent through the pipeline at all and instead the default action is applied as well.
* reduce amount of events processed in reprocess_group because of aforementioned performance penalty.
* reduce amount of attachments copied into cache, since we now know which ones are required for processing

### CODE REVIEW

File: src/sentry/reprocessing2.py [hunk 1]
```diff
@@ -208,7 +218,7 @@ def reprocess_event(project_id, event_id, start_time):
         with sentry_sdk.start_span(op="reprocess_event.set_attachment_meta"):
             attachment_cache.set(cache_key, attachments=attachment_objects, timeout=CACHE_TIMEOUT)
 
-    preprocess_event_from_reprocessing(
+    preprocess_event_from_reprocessing.delay(
```
Comment (jan-auer) [inline src/sentry/reprocessing2.py:221]: Delaying preprocessing here sounds expensive, given that we already have the event payload loaded. You could be just passing the event payload in.

**Review by jan-auer — COMMENTED** [review]
The time limit for `reprocess_group` is at `110` seconds, with a batch size of `100` events that translates to roughly one event per second. Is that a reasonable limit, given that you do quite a bit of cache and filestore access?

**Comment by untitaker** [comment]
addressed your comments, I do want to point out that the batch size has been reduced to 10 already such that the limit will not be hit (and also such that all spans are more likely below the transaction span limit of 1000)

File: src/sentry/reprocessing2.py [hunk 2]
```diff
@@ -154,27 +159,28 @@ def reprocess_event(project_id, event_id, start_time):
...
-    if data is None:
-        logger.error(
-            "reprocessing2.reprocessing_nodestore.not_found",
-            extra={"project_id": project_id, "event_id": event_id},
+    attachments = list(
+        models.EventAttachment.objects.filter(project_id=project_id, event_id=event_id)
+    )
+    files = {f.id: f for f in models.File.objects.filter(id__in=[ea.file_id for ea in attachments])}
+
+    missing_attachment_types = set(get_required_attachment_types(data)) - {
```
Comment (jjbayer) [inline src/sentry/reprocessing2.py:176]: Could `get_required_attachment_types` return a set?

Comment (untitaker) [reply src/sentry/reprocessing2.py:176]: true

**Review by jjbayer — APPROVED** [review]

File: src/sentry/reprocessing2.py [hunk 3]
```diff
@@ -154,27 +159,26 @@ def reprocess_event(project_id, event_id, start_time):
...
+        raise CannotReprocess("event.not_found")
 
-    if data is None:
-        logger.error(
-            "reprocessing2.reprocessing_nodestore.not_found",
-            extra={"project_id": project_id, "event_id": event_id},
+    attachments = list(
+        models.EventAttachment.objects.filter(project_id=project_id, event_id=event_id)
+    )
+    files = {f.id: f for f in models.File.objects.filter(id__in=[ea.file_id for ea in attachments])}
```
Comment (jan-auer) [inline src/sentry/reprocessing2.py:174]: This can be moved after `raise CannotReprocess`.

**Review by jan-auer — APPROVED** [review]

**Pull request merged**